lib/resourcemerge/core.go

```diff
@@ -266,16 +267,19 @@ func EnsureServiceType(modified *bool, existing *corev1.ServiceType, required co
 }
 
 func ensureServicePort(modified *bool, existing *corev1.ServicePort, required corev1.ServicePort) {
-	ensureServicePortDefaults(&required)
+	ensureServicePortDefaults(&required, existing)
 	if !equality.Semantic.DeepEqual(required, *existing) {
 		*modified = true
 		*existing = required
 	}
 }
 
-func ensureServicePortDefaults(servicePort *corev1.ServicePort) {
-	if servicePort.Protocol == "" {
-		servicePort.Protocol = corev1.ProtocolTCP
+func ensureServicePortDefaults(required, existing *corev1.ServicePort) {
+	if required.Protocol == "" {
+		required.Protocol = corev1.ProtocolTCP
+	}
+	if required.TargetPort == intstr.FromInt(0) || required.TargetPort == intstr.FromString("") {
+		existing.TargetPort = intstr.FromInt(int(required.Port))
```

Comment:
This defaulting is just within the required manifest; it doesn't depend on what's in-cluster at the moment.  So you can stick to a single `servicePort` argument in this function, and not pass in both `required` and `existing`.